In [1]:
import pandas as pd
import numpy as np

#### SERIES

Las series es un objeto de pandas diseñado para representar una estructura de una sola dimensión.
La series contienen dos arreglos, uno para los datos y otro para los índices. 

Para crear una serie se utiliza el constructor pd.Series().
Como podemos ver en el resultado del ejemplo, la serie tiene un índice, que es la primer columna y los datos, que es la segunda columna, y el tipo de dato de la serie es enteros.

In [2]:
a = pd.Series([9, 8, 7, 6, 5])
a

0    9
1    8
2    7
3    6
4    5
dtype: int64

Algunos atributos útiles de las series son:
- size para conocer el tamaño de la serie.
- dtype para conocer el tipo de dato de la serie.
- shape para conocer la forma de la serie, es decir, la dimención y longitud.
- index para conocer los índices, el resultado nos indica en donde inicia, donde termina y los saltos (de uno en uno, de dos en dos, etc).

Sera mediante los indices que podremos obtener y modificar los datos de la serie.

In [3]:
a.size

5

In [4]:
a.dtype

dtype('int64')

In [5]:
a.shape

(5,)

In [6]:
a.index

RangeIndex(start=0, stop=5, step=1)

Para obtener el dato de un índice basta con indicar el nombre de la serie y entre [] el indice que queremos obtener.

In [7]:
a[3]

6

Para modificar un dato de la serie basta con indicar el nombre de la serie, entre [] el indice que queremos modificar y el nuevo valor del indice.

In [8]:
a[3] = 100
a

0      9
1      8
2      7
3    100
4      5
dtype: int64

Si queremos definir nuestros propios índices, basta con indicar el parámetro index y pasarle una lista con los nombres de los índices al momento de crear la serie.
Los indices se asignan con respecto a los valores.

De igual manera, si queremos consultar o modificar un dato, debemos utilizar el nombre del índice, tal cual lo hicimos en el ejemplo anterior.

Tambien podemos asignar un nombre a la serie mediante el parámetro name y definir el tipo de dato mediante el parámetro dtype, los tipos de datos pueden ser int, float, str, bool, etc.

In [9]:
b = pd.Series([1, 2, 3, 4, 5], index=['a', 'b', 'c', 'd', 'e'], name='Numeros', dtype=int)
b

a    1
b    2
c    3
d    4
e    5
Name: Numeros, dtype: int64

### Crear series

Para crear una serie, podemos hacerlo por medio de un diccionario, donde las llaves serán los índices y los valores serán los datos de la serie.

Bata con crear o tener creado un diccionario y pasarlo como parámetro al constructor pd.Series().

In [10]:
colores = {
    'rojo' : 1,
    'azul' : 2,
    'purpura' : 3,
    'blanco' : 4,
    'negro' : 5
}

y = pd.Series(colores)
y

rojo       1
azul       2
purpura    3
blanco     4
negro      5
dtype: int64

### Valores nulos

Como sabemos, en python, los valores nulos se representan con None, pero en pandas, los valores nulos se representan con Nan(Not a numbre). 
Numpy cuenta con una constante llamada np.nan, es por eso que importamos numpy, para poder utilizar dicha constante, pero esto solo es si queremos crear una serie con valores nulos, o sea que no es necesario en caso de ya tengamos los valores en un origen definido.

In [11]:
r = pd.Series([1, np.nan, np.nan, 5, 8, 9, np.nan])
r

0    1.0
1    NaN
2    NaN
3    5.0
4    8.0
5    9.0
6    NaN
dtype: float64

El metodo isnull() nos da como resultado una nueva serie con valores boleanos, donde True indica que el valor es nulo y False que el valor no es nulo. 

In [12]:
r.isnull()

0    False
1     True
2     True
3    False
4    False
5    False
6     True
dtype: bool

Tambien podemos usar notnull(), la diferencia con isnull() es que notnull() nos da True si el valor no es nulo y False si el valor es nulo.

In [13]:
r.notnull()

0     True
1    False
2    False
3     True
4     True
5     True
6    False
dtype: bool

Si solo queremos conocer los indices con valores nulos, podemos usar la serie original y entre corchetes usar el resultado de isnull().

In [14]:
r[ r.isnull() ]

1   NaN
2   NaN
6   NaN
dtype: float64

En caso contrario de isnull(), si queremos conocer los indices con valores no nulos, usamos notnull().

In [15]:
r[ r.notnull() ]

0    1.0
3    5.0
4    8.0
5    9.0
dtype: float64

#### DataFrame

Un dataframe es una estructura de datos de pandas muy similar a las hojas de cálculo de excel o a las tablas de sql, ya que están conformadas por columnas y filas.

Para crear un dataframe, basta con crear o tener creado un diccionario.
Las llaves del diccionario serán los nombres de las columnas y los valores serán las filas (celdas con datos).

In [16]:
usuarios = {
    'username' : ['user1', 'user2', 'user3'],
    'email' : ['user1@example.com', 'user2@example.com', 'user3@example.com'],
    'age' : [25, 30, 45],
    'status' : [True, True, False]
}

l = pd.DataFrame(usuarios)
l

username              email  age  status
0    user1  user1@example.com   25    True
1    user2  user2@example.com   30    True
2    user3  user3@example.com   45   False

La creacion de un dataframe nos da una lista de indices por defecto, en el ejemplo anterior los indices son 0, 1 y 2, pero podemos definir nuestros propios indices mediante el parámetro index y pasarle una lista con los nombres de los indices.

In [17]:
l = pd.DataFrame(usuarios, index=['a', 'b', 'c'])
l

username              email  age  status
a    user1  user1@example.com   25    True
b    user2  user2@example.com   30    True
c    user3  user3@example.com   45   False

De la siguiente manera podemos conocer los datos de una columna, basta con indicar el nombre del dataframe y entre corchetes el nombre de la columna a consultar.

In [18]:
l['username']

a    user1
b    user2
c    user3
Name: username, dtype: object

Así mismo, podemos obtener el dato de una columna columna con un índice en especifico, basta con indicar en un segundo corchetes el nombre del índice que queremos consultar.

In [19]:
l['username']['a']

'user1'

Otra manera de conocer los datos de una columna, es mediante los atributos del dataframe, basta con indicar el nombre del dataframe, un punto y el nombre de la columna.
Aunque lo habitual es hacerlo de la primer manera que se mostro, es decir, mediante corchetes.

In [20]:
l.age

a    25
b    30
c    45
Name: age, dtype: int64

Si queremos consultar los nombres de las columnas, podemos usar el atributo columns del dataframe.
Basta con indicar el nombre del dataframe, un punto y columns.

In [21]:
l.columns

Index(['username', 'email', 'age', 'status'], dtype='object')

Si queremos conocer los valores que conforman nuestro dataframe, podemos usar el atributo values del dataframe. Basta con indicar el nombre del dataframe, un punto y values.

In [22]:
l.values

array([['user1', 'user1@example.com', 25, True],
       ['user2', 'user2@example.com', 30, True],
       ['user3', 'user3@example.com', 45, False]], dtype=object)

### Columnas del DataFrame


Lo primero que háremos es añadir una nueva columna al dataframe, pero antes, recordemos que las columnas son series, por lo tanto, antes de añadir la nueva columna debemos crear una serie.
En este caso, crearemos una serie don números aleatorios enteros del 5 al 10 y como nuestro dataframe solo tiene 3 filas, el tamaños de la serie también será de 3.

In [23]:
calificaciones = pd.Series( np.random.randint(5, 10, 3))
calificaciones 

0    8
1    7
2    8
dtype: int64

Ahora, para agregar la nueva columna al dataframe, basta con indicar el nombre del dataframe, entre corchetes el nombre de la nueva columna y asignarle la serie que acabamos de crear.
Pero como podemos observar, los índices no coinciden, ya que la serie que creamos tiene indices del 0 al 2 y el dataframe tiene índices a, b y c, es por eso que los valores de la columna que agregamos no aparecen y en su lugar aparecen valores nulos (NaN).

In [18]:
l['calificaciones'] = calificaciones
l

username              email  age  status  calificaciones
a    user1  user1@example.com   25    True             NaN
b    user2  user2@example.com   30    True             NaN
c    user3  user3@example.com   45   False             NaN

Para solucionar lo de los valores nulos, debemos indicar los mismos índices del dataframe al crear la serie. Posteriormente seguimos el mismo procedimiento para agregar la nueva columna al dataframe.

In [28]:
calificaciones = pd.Series( np.random.randint(5, 10, 3), index=['a', 'b', 'c'])
calificaciones 

a    7
b    6
c    9
dtype: int64

In [29]:
l['calificaciones'] = calificaciones
l

username              email  age  status  calificaciones
a    user1  user1@example.com   25    True               7
b    user2  user2@example.com   30    True               6
c    user3  user3@example.com   45   False               9

Para renombrar una columna, basta con usar el método rename() del dataframe, dentro de los paréntesis usamos el parámetro columns y le pasamos un diccionario donde la llave es el nombre actual de la columna y el valor es el nuevo nombre de la columna. Se puede hacer con una o mas columnas a la vez.

In [26]:
l.rename(
    columns={'calificaciones': 'score'}
)

username              email  age  status  score
a    user1  user1@example.com   25    True      8
b    user2  user2@example.com   30    True      6
c    user3  user3@example.com   45   False      5

Para eliminar una columna, basta con usar la palabra reservada del, indicar el nombre del dataframe y entre corchetes el nombre de la columna a eliminar. 

En este caso, pudimos eliminar la columna utilizando el nombre calificaciones aun despues de haberla renombrado, esto es porque el método rename() no modifica el DataFrame original, sino que devuelve una nueva copia con los cambios.

In [27]:
del l['calificaciones']
l

username              email  age  status
a    user1  user1@example.com   25    True
b    user2  user2@example.com   30    True
c    user3  user3@example.com   45   False

Si queremos que el cambio de nombre de una columna se aplique permanentemente al dataframe original, debemos usar el parámetro inplace=True dentro del método rename().

In [34]:
l.rename(columns={'calificaciones': 'score'}, inplace=True)
l

username              email  age  status  score
a    user1  user1@example.com   25    True      7
b    user2  user2@example.com   30    True      6
c    user3  user3@example.com   45   False      9

Al haber hecho eso, entonces si debemos escribir el nuevo nombre de la columna para que pueda ser eliminada. 

In [35]:
del l['score']
l

username              email  age  status
a    user1  user1@example.com   25    True
b    user2  user2@example.com   30    True
c    user3  user3@example.com   45   False

### Leer archivos CSV


Para poder trabajar con archivos CSV en Pandas, primero debemos importar el archivo utilizando la funcion pd.read_csv() y dentro de los paréntesis debemos colocar la ruta del archivo que queremos importar o su nombre en caso de que se encuentre en la misma carpeta que nuestro script.

El dataframe creado a partir del archivo csv se puede guardar en una variable, en este caso y, esto hace que podamos trabajar con los datos de manera más sencilla y eficiente.

Tambien incluimos el parámetro index_col='id' para indicarle a Pandas que la columna id debe ser usada como índice del dataframe. Si no hacemos esto, Pandas creará los indices numéricos por defecto, como lo habíamos visto antes. Podemos elegir cualquier columna como índice, pero lo recomendable y más comun es que se use el id.

Nota: el resultado del script nos muestra el dataframe, pero solo nos muestra las primeras 5 filas y las ultimas 5 filas de dataframe.

In [2]:
y = pd.read_csv('Breast_cancer_dataset.csv', index_col='id')
y

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
...                       ...            ...  ...            ...   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   
92751                 0.00000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0           0.16220            0.66560   
842517             158.80      1956.0           0.12380            0.18660   
84300903           152.50      1709.0           0.14440            0.42450   
84348301            98.87       567.7           0.20980            0.86630   
84358402           152.20      1575.0           0.13740            0.20500   
...                   ...         ...               ...                ...   
926424             166.10      2027.0           0.14100            0.21130   
926682             155.00      1731.0           0.11660            0.19220   
926954             126.70      1124.0           0.11390            0.30940   
927241             184.60      1821.0           0.16500            0.86810   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

Para consultar las primeras filas de nuestro dataframe, podemos usar el atributo head(), el cual sin argumentos dentro de los parentesis nos arroja las 5 primeras filas.

In [10]:
y.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760          0.3001   
842517            0.08474           0.07864          0.0869   
84300903          0.10960           0.15990          0.1974   
84348301          0.14250           0.28390          0.2414   
84358402          0.10030           0.13280          0.1980   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0            0.1622             0.6656   
842517             158.80      1956.0            0.1238             0.1866   
84300903           152.50      1709.0            0.1444             0.4245   
84348301            98.87       567.7            0.2098             0.8663   
84358402           152.20      1575.0            0.1374             0.2050   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504                0.2430          0.3613   
84348301           0.6869                0.2575          0.6638   
84358402           0.4000                0.1625          0.2364   

          fractal_dimension_worst  Unnamed: 32  
id                                              
842302                    0.11890          NaN  
842517                    0.08902          NaN  
84300903                  0.08758          NaN  
84348301                  0.17300          NaN  
84358402                  0.07678          NaN  

[5 rows x 32 columns]

Si a head() le añadimos un parametro dentro de los parentesis nos arrojara la cantidad de filas correspondientes al numero señalado entre parentesis, en este caso 20. 

In [11]:
y.head(20)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
843786           M        12.45         15.70           82.57      477.1   
844359           M        18.25         19.98          119.60     1040.0   
84458202         M        13.71         20.83           90.20      577.9   
844981           M        13.00         21.82           87.50      519.8   
84501001         M        12.46         24.04           83.97      475.9   
845636           M        16.02         23.24          102.70      797.8   
84610002         M        15.78         17.89          103.60      781.0   
846226           M        19.17         24.80          132.40     1123.0   
846381           M        15.85         23.95          103.70      782.7   
84667401         M        13.73         22.61           93.60      578.3   
84799002         M        14.54         27.54           96.73      658.8   
848406           M        14.68         20.13           94.74      684.5   
84862001         M        16.13         20.68          108.10      798.8   
849014           M        19.81         22.15          130.00     1260.0   
8510426          B        13.54         14.36           87.46      566.3   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
843786            0.12780           0.17000         0.15780   
844359            0.09463           0.10900         0.11270   
84458202          0.11890           0.16450         0.09366   
844981            0.12730           0.19320         0.18590   
84501001          0.11860           0.23960         0.22730   
845636            0.08206           0.06669         0.03299   
84610002          0.09710           0.12920         0.09954   
846226            0.09740           0.24580         0.20650   
846381            0.08401           0.10020         0.09938   
84667401          0.11310           0.22930         0.21280   
84799002          0.11390           0.15950         0.16390   
848406            0.09867           0.07200         0.07395   
84862001          0.11700           0.20220         0.17220   
849014            0.09831           0.10270         0.14790   
8510426           0.09779           0.08129         0.06664   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
843786                0.08089         0.2087  ...          23.75   
844359                0.07400         0.1794  ...          27.66   
84458202              0.05985         0.2196  ...          28.14   
844981                0.09353         0.2350  ...          30.73   
84501001              0.08543         0.2030  ...          40.68   
845636                0.03323         0.1528  ...          33.88   
84610002              0.06606         0.1842  ...          27.28 

Así mismo, con tail() obtendremos las ultimas 5 filas de nuestro dataframe.

In [12]:
y.tail()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                       
926424         M        21.56         22.39          142.00     1479.0   
926682         M        20.13         28.25          131.20     1261.0   
926954         M        16.60         28.08          108.30      858.1   
927241         M        20.60         29.33          140.10     1265.0   
92751          B         7.76         24.54           47.92      181.0   

        smoothness_mean  compactness_mean  concavity_mean  \
id                                                          
926424          0.11100           0.11590         0.24390   
926682          0.09780           0.10340         0.14400   
926954          0.08455           0.10230         0.09251   
927241          0.11780           0.27700         0.35140   
92751           0.05263           0.04362         0.00000   

        concave points_mean  symmetry_mean  ...  texture_worst  \
id                                          ...                  
926424              0.13890         0.1726  ...          26.40   
926682              0.09791         0.1752  ...          38.25   
926954              0.05302         0.1590  ...          34.12   
927241              0.15200         0.2397  ...          39.42   
92751               0.00000         0.1587  ...          30.37   

        perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                         
926424           166.10      2027.0           0.14100            0.21130   
926682           155.00      1731.0           0.11660            0.19220   
926954           126.70      1124.0           0.11390            0.30940   
927241           184.60      1821.0           0.16500            0.86810   
92751             59.16       268.6           0.08996            0.06444   

        concavity_worst  concave points_worst  symmetry_worst  \
id                                                              
926424           0.4107                0.2216          0.2060   
926682           0.3215                0.1628          0.2572   
926954           0.3403                0.1418          0.2218   
927241           0.9387                0.2650          0.4087   
92751            0.0000                0.0000          0.2871   

        fractal_dimension_worst  Unnamed: 32  
id                                            
926424                  0.07115          NaN  
926682                  0.06637          NaN  
926954                  0.07820          NaN  
927241                  0.12400          NaN  
92751                   0.07039          NaN  

[5 rows x 32 columns]

Y si añadimos un numero como parametro dentro de los parentesis, nos arrojara la cantidad de filas correspondientes al numero entre parentesis, solo que en este caso seran las utimas, como en el ejemplo, que seran las ultimas 20 filas.

In [13]:
y.tail(20)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                       
923465         B       10.820         24.21           68.89      361.6   
923748         B       10.860         21.48           68.51      360.5   
923780         B       11.130         22.44           71.49      378.4   
924084         B       12.770         29.43           81.35      507.9   
924342         B        9.333         21.94           59.01      264.0   
924632         B       12.880         28.92           82.50      514.3   
924934         B       10.290         27.61           65.67      321.4   
924964         B       10.160         19.59           64.73      311.7   
925236         B        9.423         27.88           59.26      271.3   
925277         B       14.590         22.68           96.39      657.1   
925291         B       11.510         23.93           74.52      403.5   
925292         B       14.050         27.15           91.38      600.4   
925311         B       11.200         29.37           70.67      386.0   
925622         M       15.220         30.62          103.40      716.9   
926125         M       20.920         25.09          143.00     1347.0   
926424         M       21.560         22.39          142.00     1479.0   
926682         M       20.130         28.25          131.20     1261.0   
926954         M       16.600         28.08          108.30      858.1   
927241         M       20.600         29.33          140.10     1265.0   
92751          B        7.760         24.54           47.92      181.0   

        smoothness_mean  compactness_mean  concavity_mean  \
id                                                          
923465          0.08192           0.06602        0.015480   
923748          0.07431           0.04227        0.000000   
923780          0.09566           0.08194        0.048240   
924084          0.08276           0.04234        0.019970   
924342          0.09240           0.05605        0.039960   
924632          0.08123           0.05824        0.061950   
924934          0.09030           0.07658        0.059990   
924964          0.10030           0.07504        0.005025   
925236          0.08123           0.04971        0.000000   
925277          0.08473           0.13300        0.102900   
925291          0.09261           0.10210        0.111200   
925292          0.09929           0.11260        0.044620   
925311          0.07449           0.03558        0.000000   
925622          0.10480           0.20870        0.255000   
926125          0.10990           0.22360        0.317400   
926424          0.11100           0.11590        0.243900   
926682          0.09780           0.10340        0.144000   
926954          0.08455           0.10230        0.092510   
927241          0.11780           0.27700        0.351400   
92751           0.05263           0.04362        0.000000   

        concave points_mean  symmetry_mean  ...  texture_worst  \
id                                          ...                  
923465              0.00816         0.1976  ...          31.45   
923748              0.00000         0.1661  ...          24.77   
923780              0.02257         0.2030  ...          28.26   
924084              0.01499         0.1539  ...          36.00   
924342              0.01282         0.1692  ...          25.05   
924632              0.02343         0.1566  ...          35.74   
924934              0.02738         0.1593  ...          34.91   
924964              0.01116         0.1791  ...          22.88   
925236              0.00000         0.1742  ...          34.24   
925277              0.03736         0.1454  ...          27.27   
925291              0.04105         0.1388  ...          37.16   
925292              0.04304         0.1537  ...          33.17   
925311              0.00000         0.1060  ...          38.30   
925622              0.09429         0.2128  .

### Limpieza de datos

El atributo dorpna() sirve para crear un nuevo dataframe excluyebdo todas las filas que carescan de algun valor. 

In [ ]:
y.dropna()

Si no queremos eliminar ls filas con los valores nulos, podemos utilizar el atributo fillna(), con el que podeños añadir un valor a las celdas de las filas que contienen valores nulos, como se muestra en el resultado en la columna Unnamed:32.

In [7]:
y.fillna('Nuevo valor')

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
...                       ...            ...  ...            ...   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   
92751                 0.00000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0           0.16220            0.66560   
842517             158.80      1956.0           0.12380            0.18660   
84300903           152.50      1709.0           0.14440            0.42450   
84348301            98.87       567.7           0.20980            0.86630   
84358402           152.20      1575.0           0.13740            0.20500   
...                   ...         ...               ...                ...   
926424             166.10      2027.0           0.14100            0.21130   
926682             155.00      1731.0           0.11660            0.19220   
926954             126.70      1124.0           0.11390            0.30940   
927241             184.60      1821.0           0.16500            0.86810   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

Lo anterior sirve muy bien, pero eso aplica para todas las filas, en las celdas que contienen valores nulos.

Si queremos agragar un valor en especifico a cada celda que este en determinada columna, lo haremos con un diccionario, en el que el nombre de la columna es la llave y el valor de la llave es el valor que se asignará a la celda en caso de que tenga un valor nulo.

In [24]:
y.fillna( {'radius_mean':'whithout radius', 'texture_mean':'whithout texture'} )

rojo       1
azul       2
purpura    3
blanco     4
negro      5
dtype: int64

### Atributo loc

In [25]:
usuarios = {
    'username' : ['user1', 'user2', 'user3'],
    'email' : ['user1@example.com', 'user2@example.com', 'user3@example.com'],
    'age' : [25, 30, 45],
    'status' : [True, True, False]
}

s = pd.DataFrame(usuarios, index=['a', 'b', 'c'])
s

username              email  age  status
a    user1  user1@example.com   25    True
b    user2  user2@example.com   30    True
c    user3  user3@example.com   45   False

Con el atributo loc podemos obtener filas con respecto a indices de tipo string.

In [26]:
s.loc['a']

username                user1
email       user1@example.com
age                        25
status                   True
Name: a, dtype: object

Podemos crear un subdataframe que contenga solo las filas que deseamos, basta con indicar en los parametros el nombre del índice donde comenzara y el nombre del índice en el que terminará
 este subdataframe.

In [6]:
s.loc['b':'c']

username              email  age  status
b    user2  user2@example.com   30    True
c    user3  user3@example.com   45   False

Tambien podemos obtener las filas con las columnas que puntualmente queremos, basta con indicar, despues de los nombre de los indices, los nombres de las columnas que necesitasmos.

In [7]:
s.loc['b':'c', ['email', 'status']]

email  status
b  user2@example.com    True
c  user3@example.com   False

### Atributo iloc

Con el atributo loc podemos obtener filas con respecto a indices de tipo entro (int).

In [14]:
y

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
...                       ...            ...  ...            ...   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   
92751                 0.00000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0           0.16220            0.66560   
842517             158.80      1956.0           0.12380            0.18660   
84300903           152.50      1709.0           0.14440            0.42450   
84348301            98.87       567.7           0.20980            0.86630   
84358402           152.20      1575.0           0.13740            0.20500   
...                   ...         ...               ...                ...   
926424             166.10      2027.0           0.14100            0.21130   
926682             155.00      1731.0           0.11660            0.19220   
926954             126.70      1124.0           0.11390            0.30940   
927241             184.60      1821.0           0.16500            0.86810   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

In [18]:
y.iloc[568]

diagnosis                         B
radius_mean                    7.76
texture_mean                  24.54
perimeter_mean                47.92
area_mean                     181.0
smoothness_mean             0.05263
compactness_mean            0.04362
concavity_mean                  0.0
concave points_mean             0.0
symmetry_mean                0.1587
fractal_dimension_mean      0.05884
radius_se                    0.3857
texture_se                    1.428
perimeter_se                  2.548
area_se                       19.15
smoothness_se              0.007189
compactness_se              0.00466
concavity_se                    0.0
concave points_se               0.0
symmetry_se                 0.02676
fractal_dimension_se       0.002783
radius_worst                  9.456
texture_worst                 30.37
perimeter_worst               59.16
area_worst                    268.6
smoothness_worst            0.08996
compactness_worst           0.06444
concavity_worst             

Podemos crear un subdataframe extrayendo las filas y la cantidad de filas que necesitasmos. Basta con indicar en los parametro en numero del indice de la fila en que querremos comenzar y el numero del índice de la fila en la que querremos que termine.

In [19]:
y.iloc[0:5]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760          0.3001   
842517            0.08474           0.07864          0.0869   
84300903          0.10960           0.15990          0.1974   
84348301          0.14250           0.28390          0.2414   
84358402          0.10030           0.13280          0.1980   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0            0.1622             0.6656   
842517             158.80      1956.0            0.1238             0.1866   
84300903           152.50      1709.0            0.1444             0.4245   
84348301            98.87       567.7            0.2098             0.8663   
84358402           152.20      1575.0            0.1374             0.2050   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504                0.2430          0.3613   
84348301           0.6869                0.2575          0.6638   
84358402           0.4000                0.1625          0.2364   

          fractal_dimension_worst  Unnamed: 32  
id                                              
842302                    0.11890          NaN  
842517                    0.08902          NaN  
84300903                  0.08758          NaN  
84348301                  0.17300          NaN  
84358402                  0.07678          NaN  

[5 rows x 32 columns]

Podemos crear un subdataframe con las filas y columnas que deseamos, basta con indicar, despues de los indices de la filas, los indices de las columnas que deseamos.

Nota: En el caso de iloc, al requerir de información de columnas puntuales, debemos indicar las columnas no por su nombre sino por su numero de índice, tomando en cuenta que la primer columna tiene el íncide 0, la segunda columna el índice 1, la tercer columna el índice 2 y así, sucesivamente. 

In [21]:
y.iloc[0:5, [0, 3, 6, 10]]

diagnosis  perimeter_mean  compactness_mean  fractal_dimension_mean
id                                                                          
842302           M          122.80           0.27760                 0.07871
842517           M          132.90           0.07864                 0.05667
84300903         M          130.00           0.15990                 0.05999
84348301         M           77.58           0.28390                 0.09744
84358402         M          135.10           0.13280                 0.05883

Otra forma de crear subdataframes eligiendo puntualmente las filas y columnas es de la siguiente manera.

Aunque ambas nos darán los mismos resultados, lo mas recomendable es usar esta segunda forma, ya que es mas legible y por o tanto, mas comprensible.

In [22]:
y.iloc[0:5][ ['diagnosis', 'fractal_dimension_mean'] ]

diagnosis  fractal_dimension_mean
id                                        
842302           M                 0.07871
842517           M                 0.05667
84300903         M                 0.05999
84348301         M                 0.09744
84358402         M                 0.05883

### Condicionales

In [23]:
y

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
...                       ...            ...  ...            ...   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   
92751                 0.00000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0           0.16220            0.66560   
842517             158.80      1956.0           0.12380            0.18660   
84300903           152.50      1709.0           0.14440            0.42450   
84348301            98.87       567.7           0.20980            0.86630   
84358402           152.20      1575.0           0.13740            0.20500   
...                   ...         ...               ...                ...   
926424             166.10      2027.0           0.14100            0.21130   
926682             155.00      1731.0           0.11660            0.19220   
926954             126.70      1124.0           0.11390            0.30940   
927241             184.60      1821.0           0.16500            0.86810   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

En el siguiente ejemplo, condicionamos la información que obtenemos colocando el nombre del dataframe, dentro de corchetes el nombre del dataframe, seguido de corchetes con el nombre de la columna dentro y despues el valor que queremos que tengan las filas de esa columna, en este caso 'B'.
Lo que nosa da como resultado todas las filas y columnas que sontengan B en la columna diagnosis.

In [24]:
y[ y['diagnosis'] == 'B']

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
8510426          B       13.540         14.36           87.46      566.3   
8510653          B       13.080         15.71           85.63      520.0   
8510824          B        9.504         12.44           60.34      273.9   
854941           B       13.030         18.42           82.61      523.8   
85713702         B        8.196         16.84           51.71      201.9   
...            ...          ...           ...             ...        ...   
925277           B       14.590         22.68           96.39      657.1   
925291           B       11.510         23.93           74.52      403.5   
925292           B       14.050         27.15           91.38      600.4   
925311           B       11.200         29.37           70.67      386.0   
92751            B        7.760         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
8510426           0.09779           0.08129         0.06664   
8510653           0.10750           0.12700         0.04568   
8510824           0.10240           0.06492         0.02956   
854941            0.08983           0.03766         0.02562   
85713702          0.08600           0.05943         0.01588   
...                   ...               ...             ...   
925277            0.08473           0.13300         0.10290   
925291            0.09261           0.10210         0.11120   
925292            0.09929           0.11260         0.04462   
925311            0.07449           0.03558         0.00000   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
8510426              0.047810         0.1885  ...          19.26   
8510653              0.031100         0.1967  ...          20.49   
8510824              0.020760         0.1815  ...          15.66   
854941               0.029230         0.1467  ...          22.81   
85713702             0.005917         0.1769  ...          21.96   
...                       ...            ...  ...            ...   
925277               0.037360         0.1454  ...          27.27   
925291               0.041050         0.1388  ...          37.16   
925292               0.043040         0.1537  ...          33.17   
925311               0.000000         0.1060  ...          38.30   
92751                0.000000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
8510426             99.70       711.2           0.14400            0.17730   
8510653             96.09       630.5           0.13120            0.27760   
8510824             65.13       314.9           0.13240            0.11480   
854941              84.46       545.9           0.09701            0.04619   
85713702            57.26       242.2           0.12970            0.13570   
...                   ...         ...               ...                ...   
925277             105.90       733.5           0.10260            0.31710   
925291              82.28       474.2           0.12980            0.25170   
925292             100.20       706.7           0.12410            0.22640   
925311              75.19       439.6           0.09267            0.05494   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
8510426           0.23900               0.12880          0.2977   
8510653           0.18900               0.07283          0.3184   
8510824           0.08867       

En el siguiente ajemplo solo le agregamos una condicion más al aejemplo anterior, indicando el nombre de una columna en los ultimos corchetes, esto le indica al Pandas que queremos la informacion solo de esa columna ajustandose a la primer condicional.
En este caso, solo obtendremos los radius_mean de todas filas que contienen B en la columna diagnosis.

In [25]:
y[ y['diagnosis'] == 'B']['radius_mean']

id
8510426     13.540
8510653     13.080
8510824      9.504
854941      13.030
85713702     8.196
             ...  
925277      14.590
925291      11.510
925292      14.050
925311      11.200
92751        7.760
Name: radius_mean, Length: 357, dtype: float64

En el siguiente ejemplo, obtenemos los datos de las columnas diagnosis y radius_mean pero que en la columna fractal_dimension_worst los datos solo sean mayores a 0.1000.

In [29]:
y[ y['fractal_dimension_worst'] > 0.1000][ ['diagnosis', 'radius_mean'] ]

diagnosis  radius_mean
id                             
842302           M       17.990
84348301         M       11.420
843786           M       12.450
84458202         M       13.710
844981           M       13.000
...            ...          ...
91805            B        8.571
921362           B        7.691
921386           B       14.470
925622           M       15.220
927241           M       20.600

[90 rows x 2 columns]

En el siguiente ejemplo, estamos obteniendo todas las filas donde diagnosis es B pero que en fractal_dimension_worst sea mayor a 0.1000.

Nota: para poder usar operadores logicos como & (AND) y | (OR), las conciones debemos colocarlas entre ().

In [30]:
y[ (y['diagnosis'] == 'B') & (y['fractal_dimension_worst'] > 0.1000)]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                          
859471            B        9.029         17.33           58.79      250.5   
859711            B        8.888         14.64           58.79      244.0   
8611161           B       13.340         15.86           86.49      520.0   
86409             B       14.260         19.65           97.83      629.9   
864496            B        8.726         15.83           55.84      230.9   
869476            B       11.900         14.65           78.11      432.8   
871001502         B        8.219         20.70           53.27      203.9   
8710441           B        9.731         15.34           63.78      300.2   
872608            B        9.904         18.06           64.60      302.4   
8810158           B       13.110         22.54           87.02      529.4   
883852            B       11.300         18.19           73.93      389.4   
884626            B       12.890         14.11           84.95      512.2   
894329            B        9.042         18.90           60.07      244.5   
897604            B       12.990         14.23           84.08      514.3   
901315            B       10.570         20.22           70.15      338.3   
90251             B       12.390         17.48           80.64      462.9   
903011            B       11.270         15.50           73.38      392.0   
9113239           B       13.240         20.13           86.87      542.9   
911391            B       10.880         15.62           70.41      358.9   
913063            B       12.450         16.41           82.85      476.7   
914366            B       12.650         18.17           82.69      485.6   
915186            B        9.268         12.87           61.49      248.7   
915276            B        9.676         13.14           64.12      272.5   
91544002          B       11.060         17.12           71.25      366.5   
91805             B        8.571         13.10           54.53      221.3   
921362            B        7.691         25.44           48.34      170.4   
921386            B       14.470         24.99           95.81      656.4   

           smoothness_mean  compactness_mean  concavity_mean  \
id                                                             
859471             0.10660           0.14130         0.31300   
859711             0.09783           0.15310         0.08606   
8611161            0.10780           0.15350         0.11690   
86409              0.07837           0.22330         0.30030   
864496             0.11500           0.08201         0.04132   
869476             0.11520           0.12960         0.03710   
871001502          0.09405           0.13050         0.13210   
8710441            0.10720           0.15990         0.41080   
872608             0.09699           0.12940         0.13070   
8810158            0.10020           0.14830         0.08705   
883852             0.09592           0.13250         0.15480   
884626             0.08760           0.13460         0.13740   
894329             0.09968           0.19720         0.19750   
897604             0.09462           0.09965         0.03738   
901315             0.09073           0.16600         0.22800   
90251              0.10420           0.12970         0.05892   
903011             0.08365           0.11140         0.10070   
9113239            0.08284           0.12230         0.10100   
911391             0.10070           0.10690         0.05115   
913063             0.09514           0.15110         0.15440   
914366             0.10760           0.13340         0.08017   
915186             0.16340           0.22390         0.09730   
915276             0.12550           0.22040         0.11880   
91544002           0.11940           0.10710         0.04063   
91805              0.10360           0.07632         0.02565   
921362             0.08668           0.11990    

En el siguiente ejemplo, al codigo anterior le agregamos ['fractal_dimension_worst'].mean() al final, esto para obtener el promedio de fractal_dimension_worst. 
Recordemos que ese promedio esta condicionado solo por las las filas que contienen B en la columna diagnosis y las fils con valores mayor a 0.1000 en la columna fractal_dimension_worst.

In [3]:
y[ (y['diagnosis'] == 'B') & (y['fractal_dimension_worst'] > 0.1000)]['fractal_dimension_worst'].mean()

np.float64(0.11337407407407406)

### Ordenamiento

En el siguiente ejemplo, estamos obteniendo la fila con el valor más bajo en la columna texture_mean, pero solo de las filas que contienen B en la columna diagnosis.
La primer condición en la de la columna diagnosis, que solo arroja las filas que tengan valor B en esa columna, despues ordenamos las filas con sort_values(), que ordena los valores de la columna texture_mean de menor a mayor, y por ultimo limitamos los resultados con head(1) para que solo nos arroje la primer fila, que es la que contiene el valor más bajo en la columna texture_mean.

In [4]:
y[ y['diagnosis'] == 'B'].sort_values('texture_mean').head(1)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                      
87127         B         10.8          9.71           68.77      357.6   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
id                                                                              
87127          0.09594           0.05736         0.02531              0.01698   

       symmetry_mean  ...  texture_worst  perimeter_worst  area_worst  \
id                    ...                                               
87127         0.1381  ...          12.02            73.66       414.0   

       smoothness_worst  compactness_worst  concavity_worst  \
id                                                            
87127            0.1436             0.1257           0.1047   

       concave points_worst  symmetry_worst  fractal_dimension_worst  \
id                                                                     
87127               0.04603           0.209                  0.07699   

       Unnamed: 32  
id                  
87127          NaN  

[1 rows x 32 columns]

EN el siguiente ejemplo, estamos obteniendo las 5 filas con los valores más altos en la columna texture_mean, pero solo de las filas que contienen M en la columna diagnosis.
Como podemos observar, la estructura del código es muy similar al ejemplo anterior, solo que en este caso, en sort_values() agreggamos el parámetro ascending=False, para que las filas se ordenen de mayor a menor y por ultimo limitamos los resultados con head(5) para que solo nos arroje las primeras 5 filas, que son las que contienen los 5 valores más altos en la columna texture_mean.

In [6]:
y[ y['diagnosis'] == 'M'].sort_values('texture_mean', ascending=False).head(5)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
88330202         M        17.46         39.28           113.4      920.6   
88725602         M        15.53         33.56           103.7      744.9   
88119002         M        19.53         32.47           128.0     1223.0   
88995002         M        20.73         31.12           135.7     1419.0   
925622           M        15.22         30.62           103.4      716.9   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
88330202          0.09812            0.1298          0.1417   
88725602          0.10630            0.1639          0.1751   
88119002          0.08420            0.1130          0.1145   
88995002          0.09469            0.1143          0.1367   
925622            0.10480            0.2087          0.2550   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
88330202              0.08811         0.1809  ...          44.87   
88725602              0.08399         0.2091  ...          49.54   
88119002              0.06637         0.1428  ...          45.41   
88995002              0.08646         0.1769  ...          47.16   
925622                0.09429         0.2128  ...          42.79   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
88330202            141.2      1408.0            0.1365             0.3735   
88725602            126.3      1035.0            0.1883             0.5564   
88119002            180.2      2477.0            0.1408             0.4097   
88995002            214.0      3432.0            0.1401             0.2644   
925622              128.7       915.0            0.1417             0.7917   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
88330202           0.3241                0.2066          0.2853   
88725602           0.5703                0.2014          0.3512   
88119002           0.3995                0.1625          0.2713   
88995002           0.3442                0.1659          0.2868   
925622             1.1700                0.2356          0.4089   

          fractal_dimension_worst  Unnamed: 32  
id                                              
88330202                  0.08496          NaN  
88725602                  0.12040          NaN  
88119002                  0.07568          NaN  
88995002                  0.08218          NaN  
925622                    0.14090          NaN  

[5 rows x 32 columns]

### Búsqueda por rangos

En el siguiente ejemplo, estamos obteniendo todas las filas que contienen valores en la columna texture_worst que están entre 40.00 y 50.00. Recordemos que para el uso de operadores lógicos como & (AND) y | (OR), las condiciones deben colocarse entre paréntesis.

In [8]:
y[ (y['texture_worst'] >= 40.00) & (y['texture_worst'] <= 50.00) ]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
84501001         M        12.46         24.04           83.97      475.9   
87880            M        13.81         23.75           91.56      597.8   
88119002         M        19.53         32.47          128.00     1223.0   
88203002         B        11.22         33.81           70.79      386.8   
883270           B        14.22         27.85           92.55      623.9   
88330202         M        17.46         39.28          113.40      920.6   
88725602         M        15.53         33.56          103.70      744.9   
88995002         M        20.73         31.12          135.70     1419.0   
9112085          B        13.38         30.72           86.34      557.2   
925622           M        15.22         30.62          103.40      716.9   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
84501001          0.11860           0.23960        0.227300   
87880             0.13230           0.17680        0.155800   
88119002          0.08420           0.11300        0.114500   
88203002          0.07780           0.03574        0.004967   
883270            0.08223           0.10390        0.110300   
88330202          0.09812           0.12980        0.141700   
88725602          0.10630           0.16390        0.175100   
88995002          0.09469           0.11430        0.136700   
9112085           0.09245           0.07426        0.028190   
925622            0.10480           0.20870        0.255000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
84501001             0.085430         0.2030  ...          40.68   
87880                0.091760         0.2251  ...          41.85   
88119002             0.066370         0.1428  ...          45.41   
88203002             0.006434         0.1845  ...          41.78   
883270               0.044080         0.1342  ...          40.54   
88330202             0.088110         0.1809  ...          44.87   
88725602             0.083990         0.2091  ...          49.54   
88995002             0.086460         0.1769  ...          47.16   
9112085              0.032640         0.1375  ...          41.61   
925622               0.094290         0.2128  ...          42.79   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
84501001            97.65       711.4           0.18530            1.05800   
87880              128.50      1153.0           0.22260            0.52090   
88119002           180.20      2477.0           0.14080            0.40970   
88203002            78.44       470.9           0.09994            0.06885   
883270             102.50       764.0           0.10810            0.24260   
88330202           141.20      1408.0           0.13650            0.37350   
88725602           126.30      1035.0           0.18830            0.55640   
88995002           214.00      3432.0           0.14010            0.26440   
9112085             96.69       705.6           0.11720            0.14210   
925622             128.70       915.0           0.14170            0.79170   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
84501001          1.10500               0.22100          0.4366   
87880             0.46460               0.20130          0.4432   
88119002          0.39950               0.16250          0.2713   
88203002          0.02318               0.03002          0.2911   
883270            0.30640               0.08219          0.1890   
88330202          0.32410               0.20660          0.2853   
88725602          0.57030               0.20140  

Otra manera de obtener el mismo resultado del ejemplo anterior es utilizando el método between() dentro de los corchetes, con los parametros que condicionan el rango dentro de los paréntesis.

Lo más recomendable es que, mientras puedamos, utilicemos el método between(), ya que hace que el código sea más legible y por lo tanto, más comprensible.

In [9]:
y[ y['texture_worst'].between(40.00, 50.00) ]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
84501001         M        12.46         24.04           83.97      475.9   
87880            M        13.81         23.75           91.56      597.8   
88119002         M        19.53         32.47          128.00     1223.0   
88203002         B        11.22         33.81           70.79      386.8   
883270           B        14.22         27.85           92.55      623.9   
88330202         M        17.46         39.28          113.40      920.6   
88725602         M        15.53         33.56          103.70      744.9   
88995002         M        20.73         31.12          135.70     1419.0   
9112085          B        13.38         30.72           86.34      557.2   
925622           M        15.22         30.62          103.40      716.9   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
84501001          0.11860           0.23960        0.227300   
87880             0.13230           0.17680        0.155800   
88119002          0.08420           0.11300        0.114500   
88203002          0.07780           0.03574        0.004967   
883270            0.08223           0.10390        0.110300   
88330202          0.09812           0.12980        0.141700   
88725602          0.10630           0.16390        0.175100   
88995002          0.09469           0.11430        0.136700   
9112085           0.09245           0.07426        0.028190   
925622            0.10480           0.20870        0.255000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
84501001             0.085430         0.2030  ...          40.68   
87880                0.091760         0.2251  ...          41.85   
88119002             0.066370         0.1428  ...          45.41   
88203002             0.006434         0.1845  ...          41.78   
883270               0.044080         0.1342  ...          40.54   
88330202             0.088110         0.1809  ...          44.87   
88725602             0.083990         0.2091  ...          49.54   
88995002             0.086460         0.1769  ...          47.16   
9112085              0.032640         0.1375  ...          41.61   
925622               0.094290         0.2128  ...          42.79   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
84501001            97.65       711.4           0.18530            1.05800   
87880              128.50      1153.0           0.22260            0.52090   
88119002           180.20      2477.0           0.14080            0.40970   
88203002            78.44       470.9           0.09994            0.06885   
883270             102.50       764.0           0.10810            0.24260   
88330202           141.20      1408.0           0.13650            0.37350   
88725602           126.30      1035.0           0.18830            0.55640   
88995002           214.00      3432.0           0.14010            0.26440   
9112085             96.69       705.6           0.11720            0.14210   
925622             128.70       915.0           0.14170            0.79170   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
84501001          1.10500               0.22100          0.4366   
87880             0.46460               0.20130          0.4432   
88119002          0.39950               0.16250          0.2713   
88203002          0.02318               0.03002          0.2911   
883270            0.30640               0.08219          0.1890   
88330202          0.32410               0.20660          0.2853   
88725602          0.57030               0.20140  

### Búsqueda entre opciones

En el siguiente ejemplo, estamos obteniendo todas las filas donde perimeter_mean es mayor a 100.00 y donde diagnosis sea igual a M o B.

In [10]:
y [ (y ['perimeter_mean'] > 100.00) & ( (y ['diagnosis'] == 'M') | (y ['diagnosis'] == 'B') ) ]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38           122.8     1001.0   
842517           M        20.57         17.77           132.9     1326.0   
84300903         M        19.69         21.25           130.0     1203.0   
84358402         M        20.29         14.34           135.1     1297.0   
844359           M        18.25         19.98           119.6     1040.0   
...            ...          ...           ...             ...        ...   
926125           M        20.92         25.09           143.0     1347.0   
926424           M        21.56         22.39           142.0     1479.0   
926682           M        20.13         28.25           131.2     1261.0   
926954           M        16.60         28.08           108.3      858.1   
927241           M        20.60         29.33           140.1     1265.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84358402          0.10030           0.13280         0.19800   
844359            0.09463           0.10900         0.11270   
...                   ...               ...             ...   
926125            0.10990           0.22360         0.31740   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84358402              0.10430         0.1809  ...          16.67   
844359                0.07400         0.1794  ...          27.66   
...                       ...            ...  ...            ...   
926125                0.14740         0.2149  ...          29.41   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302              184.6      2019.0            0.1622             0.6656   
842517              158.8      1956.0            0.1238             0.1866   
84300903            152.5      1709.0            0.1444             0.4245   
84358402            152.2      1575.0            0.1374             0.2050   
844359              153.2      1606.0            0.1442             0.2576   
...                   ...         ...               ...                ...   
926125              179.1      1819.0            0.1407             0.4186   
926424              166.1      2027.0            0.1410             0.2113   
926682              155.0      1731.0            0.1166             0.1922   
926954              126.7      1124.0            0.1139             0.3094   
927241              184.6      1821.0            0.1650             0.8681   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

Otra manera de obtener el mismo resultado que en el ejemplo anterior es utilizando el método isin().
Para ello primero debemos crear una lista con las opciones que queremos evaluar, en este caso 'M' y 'B', y posteriormente usamos el método isin() dentro de los corchetes, pasando como argumento el nombre de la lista que contiene las opciones. 

Este método es aplicable cuando tenemos muchas opciones que evaluar, ya que hace que el código sea más legible y por lo tanto, más comprensible.

In [11]:
diagnosis = ['M', 'B']

y [ (y ['perimeter_mean'] > 100.00) & (y ['diagnosis'].isin(diagnosis) ) ]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38           122.8     1001.0   
842517           M        20.57         17.77           132.9     1326.0   
84300903         M        19.69         21.25           130.0     1203.0   
84358402         M        20.29         14.34           135.1     1297.0   
844359           M        18.25         19.98           119.6     1040.0   
...            ...          ...           ...             ...        ...   
926125           M        20.92         25.09           143.0     1347.0   
926424           M        21.56         22.39           142.0     1479.0   
926682           M        20.13         28.25           131.2     1261.0   
926954           M        16.60         28.08           108.3      858.1   
927241           M        20.60         29.33           140.1     1265.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84358402          0.10030           0.13280         0.19800   
844359            0.09463           0.10900         0.11270   
...                   ...               ...             ...   
926125            0.10990           0.22360         0.31740   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84358402              0.10430         0.1809  ...          16.67   
844359                0.07400         0.1794  ...          27.66   
...                       ...            ...  ...            ...   
926125                0.14740         0.2149  ...          29.41   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302              184.6      2019.0            0.1622             0.6656   
842517              158.8      1956.0            0.1238             0.1866   
84300903            152.5      1709.0            0.1444             0.4245   
84358402            152.2      1575.0            0.1374             0.2050   
844359              153.2      1606.0            0.1442             0.2576   
...                   ...         ...               ...                ...   
926125              179.1      1819.0            0.1407             0.4186   
926424              166.1      2027.0            0.1410             0.2113   
926682              155.0      1731.0            0.1166             0.1922   
926954              126.7      1124.0            0.1139             0.3094   
927241              184.6      1821.0            0.1650             0.8681   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

### Métodos de string

En Pandas, podemos condicionar los valores que las series almacenen através de métodos de string.

In [16]:
usuarios = {
    'username' : ['user1', 'gabriel user2', 'user3'],
    'email' : ['auser1@example.com', 'buser2@example.com', 'auser3@example.com'],
    'age' : [25, 30, 45],
    'status' : [True, True, False]
}

d = pd.DataFrame(usuarios, index=['a', 'b', 'c'])
d

username               email  age  status
a          user1  auser1@example.com   25    True
b  gabriel user2  buser2@example.com   30    True
c          user3  auser3@example.com   45   False

Con el método str.startswith() podemos obtener todas las filas que contienen un valor en específico al inicio de la cadena de texto.
En este caso, obtenemos todas las filas donde los datos de la columna email inician con la letra 'a'.

In [18]:
d[ d['email'].str.startswith('a') ]

username               email  age  status
a    user1  auser1@example.com   25    True
c    user3  auser3@example.com   45   False

Con el método str.endswith() podemos obtener todas las filas que contienen un valor en especifico al final de la cadena de texto.
En este caso, obtenemos todas las filas donde los datos de la columna username terminan con '3'.

In [19]:
d[ d['username'].str.endswith('3') ]

username               email  age  status
c    user3  auser3@example.com   45   False

Con el método str.contains() podemos obtener todas las filas que contienen un valor en especifico en cualquier parte de la cadena de texto. 
En este caso, obtenemos todas las filas donde los datos de la columna username contienen la palabra 'gabriel'.

In [17]:
d[ d['username'].str.contains('gabriel') ]

username               email  age  status
b  gabriel user2  buser2@example.com   30    True

### Agrupamiento

Utilizando el método groupby() podemos agrupar los valores unicos de una columna. 
Basta con indicar el nombre del dataframe, groupby(), dentro de los parentesis el nombre de la columna, despues, entre corchetes accedemos a la llave por la cual se agrupa, que es el nombre de la misma columna, y posteriormente contamos con count(). 
El resultado de nuestro ejemplo es el conteo de los valores 'M' = 212  y 'B' = 357, respectivamente.

In [20]:
y.groupby('diagnosis')['diagnosis'].count()

diagnosis
B    357
M    212
Name: diagnosis, dtype: int64

En el siguiente ejemplo, condicionamos la columna perimeter_mean para que solo nos arroje las filas que contienen el valor 130.0, despues agrupamos y accedemos a la llave (ambas con el nombre de la columna diagnosis) y por ultimo contamos con count().
Esto nos da como resultado que existen 2 filas de diagnosis 'M' que contienen el valor 130.0 en la columna perimeter_mean.

In [23]:
y[ y['perimeter_mean'] == 130.0].groupby('diagnosis')['diagnosis'].count()

diagnosis
M    2
Name: diagnosis, dtype: int64